influxdb_store.go

```diff
@@ -0,0 +1,331 @@
+package appdash
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"reflect"
+	"time"
+
+	influxDBClient "github.com/influxdb/influxdb/client"
+	influxDBServer "github.com/influxdb/influxdb/cmd/influxd/run"
+	influxDBModels "github.com/influxdb/influxdb/models"
+)
+
+const (
+	dbName               string = "appdash" // InfluxDB db name.
+	spanMeasurementName  string = "spans"   // InfluxDB container name for trace spans.
+	defaultTracesPerPage int    = 10        // Default number of traces per page.
+)
+
+// Compile-time "implements" check.
+var _ interface {
+	Store
+	Queryer
+} = (*InfluxDBStore)(nil)
+
+// TODO: should be a constant.
+var zeroID = fmt.Sprintf("%016x", uint64(0))
+
+type InfluxDBStore struct {
+	con           *influxDBClient.Client // InfluxDB client connection.
+	server        *influxDBServer.Server // InfluxDB API server.
+	tracesPerPage int                    // Number of traces per page.
+}
+
+func (in *InfluxDBStore) Collect(id SpanID, anns ...Annotation) error {
+	// Current strategy is to remove existing span and save new one
+	// instead of updating the existing one.
+	// TODO: explore a more efficient alternative strategy.
+	if err := in.removeSpanIfExists(id); err != nil {
+		return err
+	}
+	// trace_id, span_id & parent_id are set as tags
+	// because InfluxDB tags are indexed & those values
+	// are uselater on queries.
+	tags := make(map[string]string, 3)
+	tags["trace_id"] = id.Trace.String()
+	tags["span_id"] = id.Span.String()
+	tags["parent_id"] = id.Parent.String()
+	// Saving annotations as InfluxDB measurement spans fields
+	// which are not indexed.
+	fields := make(map[string]interface{}, len(anns))
+	for _, ann := range anns {
+		fields[ann.Key] = string(ann.Value)
+	}
+	// InfluxDB point represents a single span.
+	pts := []influxDBClient.Point{
+		influxDBClient.Point{
+			Measurement: spanMeasurementName,
+			Tags:        tags,   // indexed metadata.
+			Fields:      fields, // non-indexed metadata.
+			Time:        time.Now(),
+			Precision:   "s",
+		},
+	}
+	bps := influxDBClient.BatchPoints{
+		Points:          pts,
+		Database:        dbName,
+		RetentionPolicy: "default",
+	}
+	_, err := in.con.Write(bps)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (in *InfluxDBStore) Trace(id ID) (*Trace, error) {
+	trace := &Trace{}
+	// GROUP BY * -> meaning group by all tags(trace_id, span_id & parent_id)
+	// grouping by all tags includes those and it's values on the query response.
+	q := fmt.Sprintf("SELECT * FROM spans WHERE trace_id='%s' GROUP BY *", id)
+	result, err := in.executeOneQuery(q)
+	if err != nil {
+		return nil, err
+	}
+	// result.Series -> A slice containing all the spans.
+	if len(result.Series) == 0 {
+		return nil, errors.New("trace not found")
+	}
+	var isRootSpan bool
+	// Iterate over series(spans) to create trace children's & set trace fields.
+	for _, s := range result.Series {
+		span, err := newSpanFromRow(&s)
```

Comment:
If possible, could you adjust the style a bit here? Prefer a blank newline after comments, like this:

```
	trace := &Trace{}

	// GROUP BY * -> meaning group by all tags(trace_id, span_id & parent_id)
	// grouping by all tags includes those and it's values on the query response.
	q := fmt.Sprintf("SELECT * FROM spans WHERE trace_id='%s' GROUP BY *", id)
	result, err := in.executeOneQuery(q)
	if err != nil {
		return nil, err
	}

	// result.Series -> A slice containing all the spans.
	if len(result.Series) == 0 {
		return nil, errors.New("trace not found")
	}
	var isRootSpan bool

	// Iterate over series(spans) to create trace children's & set trace fields.
	for _, s := range result.Series {
		span, err := newSpanFromRow(&s)
```